### PR TITLE
Build against `rdma-core` package to enable InfiniBand support

### PR DESCRIPTION
**Pull request by pentschev**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Fixes https://github.com/conda-forge/ucx-split-feedstock/issues/56

### CODE REVIEW

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:41]:
```suggestion
      run:
        - {{ pin_subpackage('rdma-core') }}
```

Note: This shouldn't be needed once a new version of `rdma-core` is packaged with `run_exports` ( https://github.com/conda-forge/rdma-core-feedstock/pull/3 )

**Comment by leofang** [comment]
It's unclear to me from the output why CUDA 10.2 + linux64 could fail, I wonder if it's the C/C++ compilers too old?

**Review by jakirkham — COMMENTED** [review]
Thanks Peter! 🙏

Added some suggestions below

It might also be worthwhile to [configure CI to retain artifacts]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#downloading-prebuilt-packages-from-ci ). This way they can be download from CI for offline testing

**Comment by jakirkham** [comment]
> It's unclear to me from the output why CUDA 10.2 + linux64 could fail, I wonder if it's the C/C++ compilers too old?

Currently `rdma-core` is building on CentOS 7 ( https://github.com/conda-forge/rdma-core-feedstock/pull/2 ). Could that be related?

**Comment by leofang** [comment]
This is what I meant, the CUDA 10.2 CI still uses gcc 7
https://github.com/conda-forge/ucx-split-feedstock/blob/89e0ca7e0286138b9a6b7fda302d2ac86962f13f/.ci_support/linux_64_c_compiler_version7cdt_namecos6cuda_compilernvcccuda_compiler_version10.2cxx_compiler_version7.yaml#L3-L4

File: recipe/meta.yaml
Comment (pentschev) [reply recipe/meta.yaml:40]: We already have a new version so I intentionally skipped this one.

Comment (jakirkham) [reply recipe/meta.yaml:40]: Would suggest keeping it since we only have a new `build/number` (not a new `version`) and conda could choose the older `build/number` package, which would leave this unconstrained

Alternatively we could try [marking the old `build/number` packages broken]( https://github.com/conda-forge/admin-requests#mark-packages-as-not-broken-on-conda-forge )

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:39]:
```suggestion
        - rdma-core  # [linux64 and cuda_compiler_version != "10.2"]
```

**Review by jakirkham — COMMENTED** [review]
Thanks Peter! 🙏

Since it sounds like we want to just try `x86_64` first, have added a suggestion below to skip other architectures

File: recipe/install_ucx.sh
Comment (jakirkham) [inline recipe/install_ucx.sh]:
```suggestion
if [[ "${target_platform}" == "linux-64" && "${cuda_compiler_version}" != "10.2" ]]; then
```

**Comment by pentschev** [comment]
All builds pass now. I've tested locally with the CUDA 11.2 package and I see the expected bandwidth:

```
$ UCX_TLS=rc,cuda_copy python -m ucp.benchmarks.send_recv -d 0 -e 1 -n 1GiB -o rmm --reuse-alloc
Server Running at 10.33.225.163:54645
Client connecting to server at 10.33.225.163:54645
Roundtrip benchmark
================================================================================
Iterations                | 10
Bytes                     | 1.00 GiB
Object type               | rmm
Reuse allocation          | True
Transfer API              | TAG
UCX_TLS                   | rc,cuda_copy
UCX_NET_DEVICES           | all
================================================================================
Device(s)                 | 0, 1
================================================================================
Bandwidth (average)       | 10.17 GiB/s
Bandwidth (median)        | 10.16 GiB/s
Latency (average)         | 98366877 ns
Latency (median)          | 98381114 ns
================================================================================
Iterations                | Bandwidth, Latency
--------------------------------------------------------------------------------
0                         | 10.16 GiB/s, 98384188ns
1                         | 10.16 GiB/s, 98378041ns
2                         | 10.17 GiB/s, 98316542ns
3                         | 10.16 GiB/s, 98392048ns
4                         | 10.16 GiB/s, 98392851ns
5                         | 10.16 GiB/s, 98386166ns
6                         | 10.16 GiB/s, 98376801ns
7                         | 10.17 GiB/s, 98303687ns
8                         | 10.16 GiB/s, 98410168ns
9                         | 10.17 GiB/s, 98328282ns
```

With the old package installed you instead see warnings such as:

```
[1684787360.984078] [dgx13:2923051:0]     ucp_context.c:1081 UCX  WARN  transport 'rc' is not available, please use one or more of: cma, cuda, cuda_copy, cuda_ipc, mm, posix, self, shm, sm, sysv, tcp
Server Running at 10.33.225.163:39161
Client connecting to server at 10.33.225.163:39161
[1684787362.517922] [dgx13:2923302:0]     ucp_context.c:1081 UCX  WARN  transport 'rc' is not available, please use one or more of: cma, cuda, cuda_copy, cuda_ipc, mm, posix, self, shm, sm, sysv, tcp
```

Also, due to [Linux's `sockaddr_un.sun_path` size limit](https://man7.org/linux/man-pages/man7/unix.7.html) (see [rdma-core conda-forge patch](https://github.com/conda-forge/rdma-core-feedstock/blob/main/recipe/0001-Remove-assertions-on-longer-than-sockaddr_un.sun_pat.patch)), we had to make sure the conda environment's path is no longer than 90 characters, if it is longer than that an error is printed and the program aborts, which is also successfully reproducible with the package:

```
$ UCX_TLS=rc,cuda_copy python -m ucp.benchmarks.send_recv -d 0 -e 1 -n 1GiB -o rmm --reuse-alloc
Server Running at 10.33.225.163:41048
Client connecting to server at 10.33.225.163:41048
path for conda-forge's environment with rdma-core cannot be longer than 90 characters
```

We would like to merge this by tomorrow if possible, so we appreciate if people here can comment, especially if there are any objections. Thanks!

**Comment by jakirkham** [comment]
@beckermr would you be interested in testing as well? No worries if not

**Comment by beckermr** [comment]
I don't think I could provide a usable test at the moment.

**Comment by jakirkham** [comment]
All good. Thought maybe you were using this somewhere given the somewhat recent issue you raised ( https://github.com/conda-forge/ucx-split-feedstock/issues/120 )

If it helps, [`ucx` can be used by `openmpi`]( https://github.com/conda-forge-admin/openmpi-feedstock/blob/a0bc3d3e127be28887ba4e0a9860cbe94e1f4a7a/recipe/meta.yaml#L47 ). So perhaps there is an MPI context where this is relevant

**Comment by leofang** [comment]
@jakirkham @pentschev I think this is not a simple rebuild, as it changes the host/run dependencies, and we should bump the build number?

**Comment by pentschev** [comment]
> @jakirkham @pentschev I think this is not a simple rebuild, as it changes the host/run dependencies, and we should bump the build number?

What is the build number in this context? Isn't that what's already being bumped in https://github.com/conda-forge/ucx-split-feedstock/pull/125/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR2 ?

**Review by leofang — APPROVED** [review]
Ah, yes, sorry I missed it for some reason. LGTM then.

**Comment by pentschev** [comment]
Thanks @jakirkham and @leofang for the help getting this through! 😄 

cc @kkraus14 for vis who may be interested in this too.

**Comment by jakirkham** [comment]
Thanks Peter! 🙏

**Pull request merged**
